include/openthread/platform/psa.h

```diff
@@ -0,0 +1,237 @@
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file includes the platform abstraction for PSA Crypto operations.
+ */
+
+#ifndef OPENTHREAD_PLATFORM_PSA_H_
+#define OPENTHREAD_PLATFORM_PSA_H_
+
+#if OPENTHREAD_CONFIG_PSA_CRYPTO_ENABLE
+
+#include <stdint.h>
+
+#include <psa/crypto_types.h>
+#include <openthread/error.h>
+#include "psa/crypto.h"
+#include "psa/crypto_values.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup plat-psa
+ *
+ * @brief
+ *   This module includes the platform abstraction for PSA Crypto.
+ *
+ * @{
+ *
+ */
+
+/**
+ * Initialise the PSA module.
+ *
+ * @retval OT_ERROR_NONE          Successfully encrypted  @p aInput.
+ * @retval OT_ERROR_FAILED        Failed to encrypt @p aInput.
+ *
+ */
+otError otPlatPsaInit(void);
+
+/**
+ * Encrypt the given data using AES-ECB cipher.
+ *
+ * @param[in]   aKeyId           Reference to the key to be used for crypto operations.
+ * @param[out]  aInput            A pointer to the input buffer.  Must not be NULL.
+ * @param[in]   aOutput           A pointer to the output buffer.  Must not be NULL.
+ *
+ * @retval OT_ERROR_NONE          Successfully encrypted  @p aInput.
+ * @retval OT_ERROR_FAILED        Failed to encrypt @p aInput.
+ * @retval OT_ERROR_INVALID_ARGS  @p aInput or @p aOutput was set to NULL.
+ *
+ */
+otError otPlatPsaEcbEncrypt(psa_key_id_t aKeyId, const uint8_t *aInput, uint8_t *aOutput);
```

Comment:
What is the reason to have otPlatPsaEcbEncrypt API? Can't we just use the official PSA API - https://github.com/ARMmbed/mbedtls/blob/v2.25.0/include/psa/crypto.h#L2113

Reply:
This is again a attempt to abstract the Crypto functionality to fit the vendors implementation needs.

Reply:
I'm fine with giving the possibility for vendors to implement things differently. However, could you help me to understand why sometimes we need `otPlat` function e.g. `otPlatPsaEcbEncrypt`, and sometimes we use direct PSA API like `psa_mac_update` or `psa_key_derivation..`. We could analogically call `psa_cipher_encrypt` for ECB. If it makes sense to create a crypto shim layer, why implement it partially?

Reply:
Yes, the approach needs to be same across the crypto modules. I will update the mac and other modules accordingly to use the abstracted APIs.